tests/pytests/unit/states/test_restconf.py

```diff
@@ -0,0 +1,164 @@
+@pytest.fixture
+def mocking_dunder_opts_test_mode_false():
+    with patch.dict(restconf.__opts__, {"test": False}) as dunder_mock:
+        yield dunder_mock
+
+
+@pytest.fixture
+def fake_uri_response_primary_blank():
+    fake_uri_response_primary_blank = [
+        True,
+        {
+            "uri_used": "primary",
+            "request_uri": "restconf/fakepath",
+            "request_restponse": {},
+        },
+    ]
+    yield fake_uri_response_primary_blank
+
+
+def test_config_manage_nochanges_testmode(
+    mocking_dunder_opts_test_mode_true, fake_uri_response_primary_blank
+):
+    with patch.dict(
+        restconf.__salt__,
+        {"restconf.uri_check": MagicMock(return_value=fake_uri_response_primary_blank)},
+    ):
+        result = restconf.config_manage("name", "restconf/test", "POST", OrderedDict())
+        assert result["result"] is True
+
+
+def test_config_manage_nochanges_realmode(
+    mocking_dunder_opts_test_mode_false, fake_uri_response_primary_blank
+):
+    with patch.dict(
+        restconf.__salt__,
+        {"restconf.uri_check": MagicMock(return_value=fake_uri_response_primary_blank)},
+    ):
+        result = restconf.config_manage("name", "restconf/test", "POST", OrderedDict())
+        assert result["result"] is True
+
+
+def test_config_manage_haschanges_testmode(
+    mocking_dunder_opts_test_mode_true, fake_uri_response_primary_blank
+):
+    with patch.dict(
+        restconf.__salt__,
+        {"restconf.uri_check": MagicMock(return_value=fake_uri_response_primary_blank)},
+    ):
+        fake_changes = OrderedDict()
+        fake_changes["fjord"] = "meow"
+        result = restconf.config_manage("name", "restconf/test", "POST", fake_changes)
+
+        assert result["result"] is None
+        assert "fjord" in str(result["changes"])
+        assert result["comment"] == "Config will be added"
+
+
+@pytest.fixture
+def mocking_dunder_salt_restconf_setdata_response_404():
+    fake_restconf_response = dict(status=404)
+    with patch.dict(
+        restconf.__salt__,
+        {"restconf.set_data": MagicMock(return_value=fake_restconf_response)},
+    ) as opt_dunder_mock:
+        yield opt_dunder_mock
+
+
+def test_config_manage_haschanges_realmode_404(
+    mocking_dunder_opts_test_mode_false,
+    fake_uri_response_primary_blank,
+    mocking_dunder_salt_restconf_setdata_response_404,
+):
+    with patch.dict(
+        restconf.__salt__,
+        {"restconf.uri_check": MagicMock(return_value=fake_uri_response_primary_blank)},
+    ):
+        fake_changes = OrderedDict()
+        fake_changes["fjord"] = "meow"
+        result = restconf.config_manage("name", "restconf/test", "POST", fake_changes)
+
+        assert result["result"] is False
+        assert len(result["changes"]) == 0
+        assert type(result["changes"]) is dict
+        assert "failed to add / modify config" in result["comment"]
+        assert "404" in result["comment"]
+        assert "restconf/fakepath" in result["comment"]
```

Comment:
**suggestion**: These assertions may be simplified. For instance, `assert not result["changes"]` will *basically* achieve the same thing here. It does allow for any `False-y` types, like None, [], "", and so-on, but that seems pretty reasonable to me.

Reply:
as in needs to be simplified or could be. I was trying to be as specific as possible

Reply:
Largely depends on the scenario. I prefer to be as "duck-like" as possible. That is, if it *looks* like a result dictionary, and *quacks* like a result dictionary, and *waddles* like a result dictionary, then that's good enough for me.

We can be flexible *and* specific. In other words - I just care that `result["changes"]` is Falsey. That's specifically what I care about, anything else is more rigid. Specifically for Salt, in `state.py` you've got lines like `if not ret["changes"] ...`. So unless there's code that is more strict, these assertions *should* be simplified :slightly_smiling_face:

Reply:
cleaned up